### PR TITLE
feat: index logs

### PR DESCRIPTION
**Pull request by greged93**
Adds the indexing of logs to the indexer.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -39,65 +37,24 @@ export default function transform({
...
-    // dataOffset <- calldata[3]
-    // dataLength <- calldata[4]
-    // calldataLen <- calldata[5]
-    const bytes = concatBytes(
-      ...calldata
-        .slice(6)
-        .map((x) => bigIntToBytes(BigInt(x))),
-    );
+    const store: {
+      collection: string;
```
Comment (Eikix) [inline src/main.ts:41]:
```suggestion
      collection: "transactions" | "logs";
```

File: src/main.ts [hunk 2]
```diff
@@ -39,65 +37,24 @@ export default function transform({
...
+      data: { tx: JsonRpcTx } | { log: JsonRpcLog };
```
[the remark this answers is not in the document]
Comment (Eikix) [reply src/main.ts:42]: but its not that important

File: src/main.ts (hunk 1 above)
Comment (greged93) [reply src/main.ts:41]: smort

File: src/main.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (greged93) [reply src/main.ts:42]: true true, will update to this

File: src/types/log.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
+// Transaction
+import { JsonRpcTx } from "../deps.ts";
+
+// Starknet
+import { Event, hash } from "../deps.ts";
+
+// Eth
+import { bigIntToHex } from "../deps.ts";
+
+const IGNORED_KEYS = [
```
Comment (Eikix) [inline src/types/log.ts:10]: let's add a comment for readability

File: src/types/log.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    transaction: JsonRpcTx;
+    event: Event;
+  },
+): JsonRpcLog | null {
+  // Filter out ignored events which aren't ETH logs.
+  if (IGNORED_KEYS.includes(event.keys[0])) {
+    return null;
+  }
+
+  if (event.keys.length < 1 || event.keys.length % 2 !== 1) {
```
Comment (Eikix) [inline src/types/log.ts:33]: let's add a comment here explaining why

File: src/types/log.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    console.error(`Invalid event ${event}`);
+    return null;
+  }
+
+  // The address is the first key of the event.
+  const address = bigIntToHex(BigInt(event.keys[0]));
+  const data = event.data.map((d) => BigInt(d).toString(16).padStart(2, "0"))
```
Comment (Eikix) [inline src/types/log.ts:40]: can we add a comment that shows an example of this small trick or explains the logic

File: src/types/log.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    .join("");
+  const topics: string[] = [];
+  for (let i = 1; i < event.keys.length; i += 2) {
+    // Topics are split tinto twos keys.
+    topics[Math.floor(i / 2)] = bigIntToHex(
+      BigInt(event.keys[i + 1]) << 128n + BigInt(event.keys[i]),
+    );
+  }
+  return {
+    removed: false,
```
Comment (Eikix) [inline src/types/log.ts:50]: what is this field?

File: src/types/transaction.ts [hunk 7]
```diff
@@ -1,18 +1,98 @@
...
+    receipt: TransactionReceipt;
+  },
+): JsonRpcTx | null {
+  const calldata = transaction.invokeV1?.calldata;
+  if (!calldata) {
+    console.error("No calldata");
+    return null;
+  }
+  const callArrayLen = BigInt(calldata[0]);
+  if (callArrayLen !== 1n) {
```
Comment (Eikix) [inline src/types/transaction.ts:47]: let's add a comment saying we don't support multicall for now?

File: src/main.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (greged93) [reply src/main.ts:42]: I tried this, but it's a little less lisible imo, maybe better to keep
```typescript
 const store: {
      collection: "transactions" | "logs";
      data: { tx: JsonRpcTx } | { log: JsonRpcLog };
    }[]
```
what do you think?

File: src/types/log.ts [hunk 8]
```diff
@@ -30,18 +32,25 @@ export function toEthLog(
     return null;
   }
 
+  // The event must have at least one key(since the first key is the address)
```
Comment (Eikix) [inline src/types/log.ts:35]: there needs to be a space between key and (

File: src/types/log.ts [hunk 9]
```diff
@@ -30,18 +32,25 @@ export function toEthLog(
...
+  // and an odd number of keys(since each topic is split into two keys).
```
Comment (Eikix) [inline src/types/log.ts:36]: same

File: src/types/log.ts [hunk 10]
```diff
@@ -30,18 +32,25 @@ export function toEthLog(
...
+  // https://github.com/kkrt-labs/kakarot/blob/main/src/kakarot/evm.cairo#L169
```
Comment (Eikix) [inline src/types/log.ts:37]: needs to be < > between link

File: src/types/log.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  if (IGNORED_KEYS.includes(event.keys[0])) {
+    return null;
+  }
+
+  if (event.keys.length < 1 || event.keys.length % 2 !== 1) {
+    console.error(`Invalid event ${event}`);
+    return null;
+  }
+
+  // The address is the first key of the event.
```
[the remark this answers is not in the document]
Comment (Eikix) [reply src/types/log.ts:43]: here: https://github.com/kkrt-labs/kakarot/blob/973d5a9216c0e4d0f92548a567d69c8f9594b5af/src/kakarot/evm.cairo#L174

[the remark this answers is not in the document]
Comment (greged93) [reply src/types/log.ts:43]: I figured since I have the link above, it was ok

**Pull request merged**